### ISSUE TITLE
Map rendering seems to suspend during app startup

### ISSUE THREAD

**Issue by sonora** [osmandapp/Osmand#3381]
I think it is worth investigating this further, it is the number one nuisance for users with mid-range or low-end devices, where OsmAnd really otherwise runs fine (!), but takes an unacceptably long time at startup.

This issue was not present several months ago, it is a possible regression.

Here is the experiment: Set the OsmAnd map to a certain zoom level (this will be your "start zoom"). Then kill OsmAnd, and re-launch.

**(1)** Independent of start zoom, it will take a certain time for the favorite layer, Position icon (if there is a fast enough fix), and the hillshade layer (if configured) to display, while the rest of the map is still empty (light gray). On a Samsung S4 Mini time until all this is displayed 9 sec, quite acceptable for everyday use.

**(2)** Interestingly, if your "start zoom is in the range 6-10, also the forest portions of the map will already be displayed at this time!

**(3)** Then it takes some additional 20-30 sec (!), where the map rendering receives no visible update, until all of a sudden all roads and other map featured are added rather quickly (roughly at the usual speed of rendering during normal later app operation).

So whatever takes so long in between step (2) and step (3) is the "extra time" we added at some point, which needs investigation. It must be something we somehow initialize after showing the forest in step (2).

**(4)** If the start zoom is >=10, the forest will not display after 10 seconds, so in these cases the user sees an empty (gray) map with favorites and position icons visible. (In this situation some users conclude the app does not detect the map at all, and kill the process and/or give us false feedback.)

**(5)** If the start zoom is exactly 9, the map after step (2) displays all forest areas, and the lines where roads will end up are already depicted in dark gray, before rendering seems to suspend

**(6)** If the start zoom is <=5, no forest is shown (as we do not render then), but it also takes the extra 20-30sec until the roads (or al very low zooms the national borders) are finally rendered.

The limit zooms for (4) and (5) may depend on device and screen magnifier, perhaps indicating this is merely a matter of "how much gets rendered before the interruption"

So in summary, it seems we somehow suspend / interrupt map rendering somewhere between showing the things (often up to the forest) in step 2, and showing the rest of the map. What do we do which apparently interrupts all rendering? Can we change init order here?

**Comment by sonora** [comment]
Some intermediate research:

I have investigated one of the devices affected here, it seems that the issue may not necessarily be caused by what we do at runtime, but may actually be an "aging" issue in our app memory or data folder:

On the device, I managed to cut down the app startup time from 30 to about 15 seconds (!), by simply this:
- Rename the existing OsmAnd data folder on the SD card (-> net.osmand.plus0)
- Uninstall OsmAnd
- Reinstall OsmAnd
- Change data folder to the SD card again (this creates a new folder net.osmand.plus, empty, but with fresh system files)
- Manually (via a file manager) move all old data files from the old (net.osmand.plus) folder to the new (net.osmand.plus) one.

Files moved were
- all map (`obf`) files, also `favourites.xml`
- the complete folders `backup`, `roads`, `srtm`, `tiles`, `tracks`, `wiki`
- Folder NOT moved were `fonts`, `rendering`, `voice`. Instead, the ones created by the new installation are now used.
- Files NOT moved were: `ind.cache` (a new one had been created with only 110kB instead of the old 231kB, `regions.ocbf` (an identical one had been re-created), `poi_types.xml` (no new one re-created so far?), and `hillshade.cache` (in the `tiles` folder, no new one re-created so far`?).

Re-using any of the previously not copied folders or system files did not increase the app start time again.

So for now it seems that we somewhere carry some "history" which created "aging", eventually causing significant increase in app start time. This was either remedied by the complete re-installation, or by the creation of the new data folder. Should be investigated on other devices (maybe some users can assist here?), as this may be a "creeping issue" unnecessarily deteriorating user experience over time.

**Comment by sonora** [comment]
Could anyone experiencing long app start times please check if temporarily renaming file `poi_types.xml` in the osmand data folder (like to `poi_types.xml0`) has any  effect? It seems this may be the culprit here ...

**Comment by hakuchi** [comment]
If I change the name from `poi_types.xml`, OsmAnd need the same time as if I not change the file.

If I change name from `ind.cache` the app start need double(first 15sec. and now 37sec.) time as before to start. Also a new one will be generated by app start.


If i add maps to the osmand folder after closing the app, OsmAnd need more time to start the app, because OsmAnd check on every start if there are new maps or if there are some deleted.
For Example:
First I start Osmand(15sec.)
then I delete all maps and use only the map from current view,
then I close Osmand,
then I start and close again(also close in the settings from app manager),
then I copy maps inside osmand folder from my PC or some other folder on my phone
then I open OsmAnd(30sec.)

I think the problem is, that OsmAnd check on every start if there are new maps or couture lines and this slow down the start. A normal user will not delete maps from osmand folder or add some new map.
OsmAnd should trust the data inside `ind.cache` when it starts. The search of new/deleted maps should be moved to the map manager if I open it in OsmAnd, or at the end of usage from OsmAnd(add some message to the end like "will you close OsmAnd Yes/No" and after tipping Yes a new `ind.cache` can be made).

This is what I think and I dont know if this is correct or if the "solution" is possible to implement.

**Comment by DFyson** [comment]
I get pretty consistent timings, about 12-14 seconds depending on poi_types.xml and if gpx tracks, favorites, hillshade and POI are all enabled or not.
But there is one a big distinction in that with poi_types.xml, everything renders at once whereas without it, everything renders in succession.

With poi_types.xml:
2.0s OsmAnd interface displays (blank background)
7.5s location shows up
13s everything shows up all at once (favs, hillshade, gpx track, offline map)

Renaming poi_types.xml:
2.0s OsmAnd interface displays (blank background)
9.5s location + favs
10.5s hillshade
11.5s offline map
13.5s gps track

Using 2.5.4+ from 2016-11-22 on Samsung Galaxy S3

**Comment by sonora** [comment]
Yes, I can confirm that with renamed poi_types.xml the map is finally displayed all in one go. But at least on many older devices this seems to shave some 10-20 sec off the overall time it takes (tested e.g. on Samsung S4 Mini, or an HTX Desire X).

**Comment by vshcherb** [comment]
Interesting test. Initially I was struggling to get time down to 5 seconds, unfortunately the best I achieved was 9 seconds. Then we made decision to show at least UI asap to split waiting time in chunks. So, 9-12 seconds is the vast majority of devices. Everything longer could be suspicious especially for new devices.
Deleting ind.cache could definitely increase the time up to minutes. Since poi.xml is growing the time will grow but should not be significant.

**Comment by aceman444** [comment]
I've even seen cases where a much more powerful device hanged longer on the start screen, than o slower device (tested with v2.5). v1.9 was way faster, but that is understandable, it showed less details and poi types.

But something really should be done with the startup time.

What is happening in the first 9s in the comment from @DFyson ? Is it reading parsing those xml files of poi definitions?

**Comment by stephan75** [comment]
So have a look at JOSM Editor based on Java on a desktop PC:

When starting josm up, it prints all steps of starting procedure with amount of time used in miliseconds.

Can we have a feature in development plugin of Osmand to see similar logging messages on screen or in an extra logging file?

**Comment by sonora** [comment]
Funny thing is: Devices with renamed poi_types.xml seem to function without any adverse effect, or am I overlooking something??

**Comment by vshcherb** [comment]
```Can we have a feature in development plugin of Osmand to see similar logging messages on screen or in an extra logging file?```
Yes, it is there in the log file.

**Comment by sonora** [comment]
What's the purpose of poi_types.xml, or the deficit if you rename it?

**Comment by vshcherb** [comment]
poi_types is the file with all pois, it is read from file system for speedup I think.

**Comment by sonora** [comment]
Well, looks like a "slow down", on sone devices, at least? :)

Why should it be faster anyway on the external (potentially extSdCard) file system, rather than keeping it in internal app memory if we really need it?

**Comment by Tony-il-Capo** [comment]
On my old Sony Xperia S, I see a difference between having  poi_types.xml renamed or not.
With original poi_types.xml, seems to be ~27s, and renamed ~17
And just 2s for both, if memory is not cleared before starting. 
All times measured right at the osmand icon tap and stopped when map finished rendering.

**Comment by sonora** [comment]
Exactly my experience!

(I guess your "2s" case is if the app was not killed, but is merely called to the foreground again, for me this is basically instsntaneous. But this issue means to talk about a fresh app start, and there I have many devices confirming your 27sec vs. 17sec experience, which is a lot.)

**Comment by vshcherb** [comment]
27 seconds experience is a lot. Though we checked on our devices (range of 7) it is not that dramatic 8-13 seconds. I will try to reveal old devices, though very old devices run 2.3 they are not supported :)))

**Comment by sonora** [comment]
I am talking e.g. a Samsung S 4 Mini with stock Android 4.4.2, quite a common device (27 sec). Also an HTC Desire X, with stock Android 4.1.1 (35 sec). Renaming poi_types.xml to poi_types.xml0 on both reliably reduces these times by about 10 sec each.

I do not find any issues, after renaming everything seems to function normally??

**Comment by sonora** [comment]
@vshcherb If we are not sure what the poi_types.xml file is good for at all, but it clearly improves speed on many devices to not have it present, I guess we should not write this file to the file system any more upo installation or updating, and even rename it if present- upon??

**Comment by vshcherb** [comment]
@sonora Indeed copying poi types is not required. Here is the line to comment out https://github.com/osmandapp/Osmand/blob/master/OsmAnd/src/net/osmand/plus/resources/ResourceManager.java#L554. Though it require testing first  and second there is a purpose of this.

It is the same as for routing.xml, so somebody could customize it. But if you really think that's a reason please comment it out in that case 3rd party poi_types will be supported anyway but the file won't be copied.

**Comment by sonora** [comment]
Victor, just tested this: It is **not** the copy process `copyPoiTypes()` that slows down the app start. It is the **mere presence** of the `poi_types.xml` file which slows down things.

Meaning: What improves the startup speed is renaming (or deleting) the` poi_types.xml` file. Since commenting out `copyPoiTypes()` only prevents its re-creation upon app updates, that would only fix fresh installs, not update installs (where the file already exists).

It looks like we are somehow not at the bottom of this: The issue is not writing the file itself, it is something we do when the file is present (like reading/importing/analyzing it ?), but simply skip when the file is not there. So the fix must be that we somehow automatically shortcut (skip) processing the poi_types file if it is unaltered from the default we wrote to the external file system upon app installation or update?  (99.9% of users never change the file, but experience the slowdown from it being freshly processed upon every app start.)

**Comment by sonora** [comment]
Let me re-open this so we don't forget. Still think this is worth a "proper" fix.

**Comment by vshcherb** [comment]
@sonora The above fix is proper
With it we don't write files on new device/installation, so there is no poi_types present. For old installations we can delete it manually. Deleting file automatically may not be desired in case it was customized.

**Comment by aceman444** [comment]
You can compare if the file is the same as the one inside the package.

**Comment by sonora** [comment]
I have added the "simple" fix: https://github.com/osmandapp/Osmand/commit/2740ad105df3785ed2a91f987a6c087e091af1f1

Yes,
- a comparison would be good, as suggested above,
- or at least a remark somewhere in our documentation that you can delete the file if you have never modified it to speed up the app start.

Unfortunately I will be on bad Internet links for a few weeks so cannot implement either, maybe someone can help out  :). Thx!
